Makefile

```diff
@@ -89,139 +102,104 @@ vendor: glide.yaml
 	if [ "$(LIBCALICOGO_PATH)" != "none" ]; then \
           EXTRA_DOCKER_BIND="-v $(LIBCALICOGO_PATH):/go/src/github.com/projectcalico/libcalico-go:ro"; \
 	fi; \
-  docker run --rm \
-    -v $(CURDIR):/go/src/$(PACKAGE_NAME):rw $$EXTRA_DOCKER_BIND \
-    -v $(HOME)/.glide:/home/user/.glide:rw \
-    -e LOCAL_USER_ID=$(LOCAL_USER_ID) \
-    $(GO_BUILD_CONTAINER) /bin/sh -c ' \
+    docker run --rm -i \
+      -v $(CURDIR):/go/src/$(PACKAGE_NAME):rw $$EXTRA_DOCKER_BIND \
+      -v $(HOME)/.glide:/home/user/.glide:rw \
+      -e LOCAL_USER_ID=$(LOCAL_USER_ID) \
+      $(GO_BUILD_CONTAINER) /bin/sh -c ' \
 		  cd /go/src/$(PACKAGE_NAME) && \
-      glide install -strip-vendor'
+          glide install -strip-vendor'
+
+# The supported different binary names. For each, ensure that an OS and ARCH is set
+bin/calicoctl-%-amd64: ARCH=amd64
+bin/calicoctl-%-arm64: ARCH=arm64
+bin/calicoctl-%-ppc64le: ARCH=ppc64le
+bin/calicoctl-%-s390x: ARCH=s390x
+bin/calicoctl-darwin-amd64: OS=darwin
+bin/calicoctl-windows-amd64: OS=windows
+bin/calicoctl-linux-%: OS=linux
+
+bin/calicoctl-%: $(CALICOCTL_FILES) vendor
+	mkdir -p bin
+	-mkdir -p .go-pkg-cache
+	docker run --rm -ti \
+	  -e OS=$(OS) -e ARCH=$(ARCH) \
+	  -e GOOS=$(OS) -e GOARCH=$(ARCH) \
+	  -e CALICOCTL_VERSION=$(CALICOCTL_VERSION) \
+	  -e CALICOCTL_BUILD_DATE=$(CALICOCTL_BUILD_DATE) -e CALICOCTL_GIT_REVISION=$(CALICOCTL_GIT_REVISION) \
+	  -v $(CURDIR):/go/src/$(PACKAGE_NAME):ro \
+	  -v $(CURDIR)/bin:/go/src/$(PACKAGE_NAME)/bin \
+      -e LOCAL_USER_ID=$(LOCAL_USER_ID) \
+      -v $(CURDIR)/.go-pkg-cache:/go-cache/:rw \
+      -e GOCACHE=/go-cache \
+	    $(GO_BUILD_CONTAINER) sh -c '\
+          cd /go/src/$(PACKAGE_NAME) && \
+          go build -v -o bin/calicoctl-$(OS)-$(ARCH) $(LDFLAGS) "./calicoctl/calicoctl.go"'
+
+# Overrides for the binaries that need different output names
+bin/calicoctl: bin/calicoctl-linux-amd64
+	mv $< $@
```

Comment:
I think this should be 

```
cp $< $@
```

We want to keep around `bin/calicoctl-linux-amd64` but _add_ the archless/OSless default of `bin/calicoctl`